content/advent-2016/how-to-write-good-tests-for-solid-code.md

```diff
@@ -0,0 +1,219 @@
++++
+author = ["Ernesto Jimenez"]
+date = "2016-12-16T00:00:00+00:00"
+series = ["Advent 2016"]
+title = "Writing good unit tests for SOLID Go"
++++
+
+Dave Cheney covered how interfaces are used to design good Go code in his [SOLID Go Design talk][solid-go-talk] and [blog post][solid-go-post].
+
+In this blog post, we are going to focus on some tips on how to write unit tests for that beautiful SOLID code.
+
+The primary objectives for our test builds are to:
+
+- **Increase the confidence in our code**. Otherwise, the test code is just dead weight.
+- **Be fast**. Who likes waiting 15 minutes for the tests to pass?
+- **Be stable**. Tests should never fail randomly, and a small change should never break unrelated tests.
+- **Be short**. It must be the shortest possible checks required to increase our confidence in the code.
+
+Most of us have had to suffer test suites which fell short of achieving these objectives. Objects with dependencies are, in many cases, the most problematic.
+
+## Starting with an example
+
+When we implement a SOLID design, as a rule of thumb our **structs will depend on interfaces instead of structs**.
+
+Let's look at a simplified example of Go code that would use interfaces for its dependencies.
+
+We are building a struct to manage users in our application. We will call it `UserManager`.
+
+```go
+// UseManager
+type UserManager struct {
+    notifier UserNotifier // Used to schedule emails
+    store    UserStore    // Used to persist users
+}
+
+// SignUp creates user account pending activation and sends the activation email.
+// It will return an error if creating the user failed  due to invalid details or problems with the UserStore.
+func (um *UserManager) SignUp(ctx context.Context, user User) (*User, error) {
+    // [...]
+}
+
+// [...]
+```
+
+As we can see, `UserManager` is dependent on the following two interfaces:
+
+```go
+// UserNotifier specifies the methods to schedule emails to be sent to certain user
+type UserNotifier interface {
+    RequestActivation(ctx context.Context, id string) error
+    RecoverPassword(ctx context.Context, id string) error
+}
+
+// UserStore specifies the methods required to manage user accounts
+type UserStore interface {
+    Find(ctx context.Context, id string) (*User, error)
+    Update(context.Context, User) error
+    Create(context.Context, User) (*User, error)
+}
+```
+
+Following the SOLID principles, we will create a function to initialise the `UserManager` injecting the dependencies:
+
+```go
+func NewUserManager(store UserStore, notifier UserNotifier) *UserManager {
```

Comment:
I prefer using constructors because sometimes the struct has private fields like locks or channels.

By setting the constructor from the get go even when there are no private fields to be set you are consistent in the way you initialise the structs and can later add a private field if required without having to manually change all the struct literals.